### ISSUE TITLE
Picture saving to local disk on local server install

### ISSUE THREAD

**Issue by Klimbim1** [FarmBot/Farmbot-Web-API#331]
My config/application.yml with regards to the google cloud config looks like:

```
# Google Cloud Storage API Bucket for image data. Deleting this will save to disk
##GCS_BUCKET: "GOOGLE_CLOUD_STORAGE_BUCKET_NAME_FOR_IMAGE_FILES"
# Google Cloud Storage ID for image data. Deleting this will save images to disk.
##GCS_ID: "GOOGLE_CLOUD_STORAGE: 'interop' id"
##GCS_KEY: "GOOGLE_CLOUD_STORAGE: 'interop' key"
```

If I try to take a picture it says:
![bildschirmfoto 2017-05-31 um 23 25 46](https://cloud.githubusercontent.com/assets/28235189/26654834/845efbe4-4658-11e7-92d0-653a27061760.png)

Where do I need to adapt the config?

**Comment by RickCarlino** [comment]
@Klimbim1 Please try this:

1. Run `rails jobs:work`. (This was recently added to the setup- I've updated the docs to reflect that)
2. Wait a few seconds.
3. Check the photo widget (`http://localhost:3000/app/farmware`)

Do any images come up on that page?

**If you do see an image:** you can safely ignore that log message and we will change the message in a future release. This is just a verbiage issue. You can find the real image URL by right clicking the image and selecting `copy image address`.

**If you do not see any images:** There may be another issue at play. Please send me more error logs as well as the contents of the Javascript console (CTRL + SHIFT + J) and I can take a deeper look.

**Comment by Klimbim1** [comment]
Hi @RickCarlino,
it does not work; If I execute rails jobs:work this does something, here is a log file sequence which
comes over and over:

> >  Delayed::Backend::ActiveRecord::Job Load (1.0ms)  UPDATE "delayed_jobs" SET locked_at = '2017-06-05 17:07:34.876563', locked_by = 'host:ubuntu pid:7511' WHERE id IN (SELECT  "delayed_jobs"."id" FROM "delayed_jobs" WHERE ((run_at <= '2017-06-05 17:07:34.875197' AND (locked_at IS NULL OR locked_at < '2017-06-05 13:07:34.875296') OR locked_by = 'host:ubuntu pid:7511') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNING *
> DEPRECATION WARNING: cleanup! is deprecated and will be removed from Rails 5.1 (use Rails.application.reloader.reload! instead of cleanup + prepare) (called from require at bin/rails:4)
> DEPRECATION WARNING: prepare! is deprecated and will be removed from Rails 5.1 (use Rails.application.reloader.prepare! instead) (called from require at bin/rails:4)
>   Delayed::Backend::ActiveRecord::Job Load (1.5ms)  UPDATE "delayed_jobs" SET locked_at = '2017-06-05 17:07:39.996718', locked_by = 'host:ubuntu pid:7511' WHERE id IN (SELECT  "delayed_jobs"."id" FROM "delayed_jobs" WHERE ((run_at <= '2017-06-05 17:07:39.995229' AND (locked_at IS NULL OR locked_at < '2017-06-05 13:07:39.995330') OR locked_by = 'host:ubuntu pid:7511') AND failed_at IS NULL) ORDER BY priority ASC, run_at ASC LIMIT 1 FOR UPDATE) RETURNING *
> DEPRECATION WARNING: cleanup! is deprecated and will be removed from Rails 5.1 (use Rails.application.reloader.reload! instead of cleanup + prepare) (called from require at bin/rails:4)
> DEPRECATION WARNING: prepare! is deprecated and will be removed from Rails 5.1 (use Rails.application.reloader.prepare! instead) (called from require at bin/rails:4)
>   Delayed::Backend::ActiveRecord::Job Load (1.0ms) 

Then this thingi appeared, I tried to go to the URL but it only says sorry :)

>2017-06-05T10:10:59-0700: [Worker(host:ubuntu pid:7511)] Job ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper (id=1) RUNNING

>   User Load (1.5ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 108], ["LIMIT", 1]]
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46] Performing ActionMailer::DeliveryJob from DelayedJob(mailers) with arguments: "UserMailer", "welcome_email", "deliver_now", #<GlobalID:0x0000000694a588 @uri=#<URI::GID gid://farm-bot/User/108>>
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendering layouts/mailer.html.erb
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendering user_mailer/welcome_email.html.erb within layouts/mailer
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendered user_mailer/welcome_email.html.erb within layouts/mailer (0.9ms)
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendered layouts/mailer.html.erb (2.8ms)
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendering user_mailer/welcome_email.text.erb
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46]   Rendered user_mailer/welcome_email.text.erb (0.8ms)
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46] UserMailer#welcome_email: processed outbound mail in 24.9ms
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46] Sent mail to myemailadress@myurl.com (8.5ms)
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46] Date: Mon, 05 Jun 2017 10:10:59 -0700
> From: farmbot-mailer@192.168.0.38
> To: myemailadress@myurl.com
> Message-ID: <593590a3ce468_1d57ccb0dc3172b@ubuntu.mail>
> Subject: Welcome to The FarmBot Web App!
> Mime-Version: 1.0
> Content-Type: multipart/alternative;
>  boundary="--==_mimepart_593590a3cd3f6_1d57ccb0dc3162e";
>  charset=UTF-8
> Content-Transfer-Encoding: 7bit
> 
> 
> ----==_mimepart_593590a3cd3f6_1d57ccb0dc3162e
> Content-Type: text/plain;
>  charset=UTF-8
> Content-Transfer-Encoding: 7bit
> 
> Welcome to FarmBot, myuser
> =====================================
> 
> You please verify your email by visiting http://192.168.0.38:3000/verify.html?token=ERROR
> 
> Thanks for joining and have a great day!
> 
> ----==_mimepart_593590a3cd3f6_1d57ccb0dc3162e
> Content-Type: text/html;
>  charset=UTF-8
> Content-Transfer-Encoding: 7bit
> 
> <!DOCTYPE html>
> <html>
>   <head>
>     <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
>   </head>
>   <body>
>     <h1>Welcome to FarmBot, myuser</h1>
> <p>
>   Please verify your email by clicking here
>   <a href="http://192.168.0.38:3000/verify.html?token=ERROR">here</a>.<br>
> </p>
> <p>Thanks for joining and have a great day!</p>
> 
>   </body>
> </html>
> ----==_mimepart_593590a3cd3f6_1d57ccb0dc3162e--
> 
> [ActiveJob] [ActionMailer::DeliveryJob] [f5d7909b-3775-46da-8c67-3fb863483f46] Performed ActionMailer::DeliveryJob from DelayedJob(mailers) in 44.34ms
> [Worker(host:ubuntu pid:7511)] Job ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper (id=1) FAILED (4 prior attempts) with Errno::ECONNREFUSED: Connection refused - connect(2) for "localhost" port 1025
> 2017-06-05T10:10:59-0700: [Worker(host:ubuntu pid:7511)] Job ActiveJob::QueueAdapters::DelayedJobAdapter::JobWrapper (id=1) FAILED (4 prior attempts) with Errno::ECONNREFUSED: Connection refused - connect(2) for "localhost" port 1025
>    (1.0ms)  BEGIN
>   SQL (1.3ms)  UPDATE "delayed_jobs" SET "attempts" = $1, "run_at" = $2, "locked_at" = $3, "locked_by" = $4, "updated_at" = $5 WHERE "delayed_jobs"."id" = $6  [["attempts", 5], ["run_at", 2017-06-05 17:21:29 UTC], ["locked_at", nil], ["locked_by", nil], ["updated_at", 2017-06-05 17:10:59 UTC], ["id", 1]]
>    (1.1ms)  COMMIT

If I visit this URL: http://192.168.0.38:3000/verify.html?token=ERROR 
it says in the error java console log:

> :3000/api/users/verify/ERROR Failed to load resource: the server responded with a status of 404 (Not Found)
> verification.ts:66 Uncaught (in promise) Error: USER VERIFICATION FAILED!
>     at eval (verification.ts:66)
>     at step (verification.ts:32)
>     at Object.eval [as throw] (verification.ts:13)
>     at rejected (verification.ts:5)
>     at <anonymous>

Cheers!

**Comment by RickCarlino** [comment]
@Klimbim1 May I see your `application.yml`? Self hosted setups are very different than the setup used on the main server. I suspect you are hitting an edge case that we do not test against.

**Please remove these keys (if any) before sharing:** `GCS_KEY`, `RSA_KEY`, `ACME_SECRET`, `DATABASE_URL`, `DEVISE_SECRET`, `GCS_KEY`, `ROLLBAR_ACCESS_TOKEN`, `SENDGRID_PASSWORD`.

**Comment by Klimbim1** [comment]
@RickCarlino for sure, here you are:
` ░██████░░███████░░█████░░░██████░░
# ░██░░░░░░░░██░░░░██░░░██░░██░░░██░  READ THIS BEFORE USING IT.
# ░██████░░░░██░░░░██░░░██░░██████░░  SEE NOTES BELOW:
# ░░░░░██░░░░██░░░░██░░░██░░██░░░░░░
# ░██████░░░░██░░░░░████░░░░██░░░░░░
# You will hit issues if any of these are set to the wrong value.
# Please read each line of this file before starting the server.
#
#  =============================================================================
#
# Self hosting users can safely delete this (a new key will be created).
# This key is used to exchange secrets between bots and MQTT servers (important
# if you don't use SSL)
# SERVER WONT WORK IF YOU FORGET TO DELETE THIS EXAMPLE TEXT BELOW.
# ADD A REAL RSA_KEY OR DELETE THIS LINE!!
##RSA_KEY: "Change this! Keys look like `-----BEGIN RSA .........`"
# If you use Let's Encrypt for SSL,
# you must set this when renewing SSL.
# Otherwise, not required and CAN BE REMOVED.
##ACME_SECRET: "-----"
# If your server is on a domain (eg: my-own-farmbot.com), put it here.
# 0.0.0.0 is a good value because it exposes all IPs on your machine.
# DONT USE `localhost`. USE A REAL IP!
API_HOST: "192.168.0.38"
# 3000 for local development. 443 is using SSL. You will need `sudo` to use PORT
# 80 on most systems.
API_PORT: "3000"
# This can be set to anything.
# Most users can just delete it.
# This is used for people writing modifications to the software, mostly.
##DOCS: "Set this to any value if you want to generate API docs after running tests"
# Most users can delete this.
# Used by people who pay for managed database hosting.
DATABASE_URL: XXXX
# MUST REPLACE.
# Generate a secret by typing `rake secret` into the console.
DEVISE_SECRET: XXXX
# Most personal server users can delete this.
##FORCE_SSL: "Remove this if not using HTTPS://"
# Firmware update server. Use default if you don't have a special use case.
# Off grid servers may have issues connecting to our update URL.
FW_UPDATE_SERVER: "https://api.github.com/repos/FarmBot/farmbot-arduino-firmware/releases/latest"
OS_UPDATE_SERVER: "https://api.github.com/repos/farmbot/farmbot_os/releases/latest"
# Google Cloud Storage API Bucket for image data.
# Deleting this will save to disk.
# Most self hosting users will want to delete this.
##GCS_BUCKET: "GOOGLE_CLOUD_STORAGE_BUCKET_NAME_FOR_IMAGE_FILES"
# Google Cloud Storage ID for image data.
# Deleting this will save images to disk.
# Most self hosting users will want to delete this.
##GCS_ID: "GOOGLE_CLOUD_STORAGE: 'interop' id"
# Most self hosting users will want to delete this.
##GCS_KEY: "GOOGLE_CLOUD_STORAGE: 'interop' key"
# Most self hosting users will want to delete this.
##HEROKU_SLUG_COMMIT: "This is set by Heroku, used by Frontend to show current version."
# Where is your MQTT server running?
# Use a REAL IP ADDRESS if you are controlling real bots.
# 0.0.0.0 is only OK for software testing. Change this!
MQTT_HOST: "192.168.0.38"
# Delete this line if you are not an employee of FarmBot, Inc.
##NPM_ADDON: "Used by FarmBot, Inc. to load proprietary extras, like Rollbar."
# Same as above. Can be deleted unless you are a Rollbar.IO customer.
##ROLLBAR_ACCESS_TOKEN: "____"
# FarmBot, Inc. uses SendGrid to send emails.
# Delete these if you don't use send grid.
##SENDGRID_PASSWORD: "Used by FarmBot, Inc"
##SENDGRID_USERNAME: "Used by FarmBot, Inc"
# For email delivery. Who is your email host?
SMTP_HOST: "smtp.sendgrid.net"
# Optional with default of 587
SMTP_PORT: 587
# Set this if you don't want to deal with email verification of new users.
# (self hosted users)
NO_EMAILS: TRUE`

**Comment by RickCarlino** [comment]
Please try the following:

1. Pull down the latest version: `git pull origin master`
2. Update dependencies: `bundle install`
3. Migrate your database: `rails db:migrate`
4. Install the latest frontend: `./install_frontend.sh`
5. Start the server

and try taking a photo again. Do you see anything on the `farmware` page? Do you see any errors?

I found a few issues that might affect folks who do not have an email server, but I am not sure it was related to anything here.

Also, could you try running `rspec spec`? You should see `0 failures` after running the tests.

**Comment by Klimbim1** [comment]
Hi @RickCarlino, 
@MrChristofferson recommended to me to fork all three projects, mqtt, frontend and web-api in order to be up to date and see the changes on a lets say weekly basis when I pull the latest versions.

Therefore I need to adapt the frontend starting thing. Could you elaborate how to correctly start the frontend?

However if I go into Farmbot-Web-API and type rspec spec it says the following:
myuser:~/Farmbot-Web-API$ rspec spec
WARNING: Skipping key "FW_UPDATE_SERVER". Already set in ENV.
WARNING: Skipping key "OS_UPDATE_SERVER". Already set in ENV.
WARNING: Skipping key "MQTT_HOST". Already set in ENV.
WARNING: Use strings for Figaro configuration. 587 was converted to "587".
WARNING: Use strings for Figaro configuration. true was converted to "true".

Randomized with seed 25127
F

Failures:

  1) Api::ToolsController#create creates a new tool
     Failure/Error: let(:user) { FactoryGirl.create(:user) }
     
     NoMethodError:
       undefined method `password=' for #<User:0x0000000737c358>
     # ./spec/controllers/api/tools/create_spec.rb:6:in `block (3 levels) in <top (required)>'
     # ./spec/controllers/api/tools/create_spec.rb:10:in `block (3 levels) in <top (required)>'

Finished in 0.08596 seconds (files took 8.45 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/controllers/api/tools/create_spec.rb:9 # Api::ToolsController#create creates a new tool

Randomized with seed 25127

Coverage report generated for RSpec to /home/klimbim/Farmbot-Web-API/coverage. 1174 / 1833 LOC (64.05%) covered.

I looked at the install_frontend.sh and I am basically stuck with 
myuser:~/farmbot-web-frontend$ npm run build

it says at the end:
`
ERROR in /home/myuser/farmbot-web-frontend/src/images/__tests__/image_flipper_test.tsx
(5,25): error TS7016: Could not find a declaration file for module 'deep-freeze'. '/home/myuser/farmbot-web-frontend/node_modules/deep-freeze/index.js' implicitly has an 'any' type.
Child extract-text-webpack-plugin:
                                   Asset     Size  Chunks             Chunk Names
    3f0fc622e73ad38dc3e0bbb86eb73182.eot  77.5 kB          [emitted]  
    95a702087e984248205235ac8cfe1c43.eot  79.4 kB          [emitted]  
       [0] ./~/css-loader/lib/css-base.js 1.51 kB {0} [built]
       [1] ./~/css-loader!./~/@blueprintjs/core/dist/blueprint.css 178 kB {0} [built]
       [2] ./~/css-loader!./~/sass-loader!./src/css/_index.scss 43.9 kB {0} [built]
       [3] ./~/@blueprintjs/core/resources/icons/icons-16.eot 82 bytes {0} [built]
       [4] ./~/@blueprintjs/core/resources/icons/icons-20.eot 82 bytes {0} [built]
       [5] ./~/@blueprintjs/core/resources/icons/icons-16.ttf 103 kB {0} [built]
       [6] ./~/@blueprintjs/core/resources/icons/icons-16.woff 103 kB {0} [built]
       [7] ./~/@blueprintjs/core/resources/icons/icons-20.ttf 106 kB {0} [built]
       [8] ./~/@blueprintjs/core/resources/icons/icons-20.woff 106 kB {0} [built]
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! farmbot-web-frontend@1.1.0 build: `node_modules/webpack/bin/webpack.js --config tools/webpack.config.prd.js --display-error-details`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the farmbot-web-frontend@1.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/myuser/.npm/_logs/2017-06-07T20_09_10_499Z-debug.log`

I tried to install the deep-freeze module but that does not work either... :/ any ideas?

**Comment by RickCarlino** [comment]
@Klimbim1 I saw that error yesterday and thought I patched it already :thinking: .

What do you see when you run `git log -1`?

**Comment by Klimbim1** [comment]
```
/Farmbot-Web-API$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	deleted:    config/application.example.yml
	deleted:    config/database.example.yml
	modified:   db/schema.rb
	modified:   package.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	package-lock.json

no changes added to commit (use "git add" and/or "git commit -a")

```
`:~/mqtt-gateway$ git log -1
commit 63a655deb8e3b2f0e5327389bff6256554d67098
Merge: 60f5f79 0f4c66b
Author: Rick Carlino <RickCarlino@users.noreply.github.com>
Date:   Tue Apr 25 07:13:14 2017 -0500

    Merge pull request #6 from FarmBot/roryaronson-license-patch
    
`

**Comment by RickCarlino** [comment]
@Klimbim1 Your local version is two months behind master. As such, it will not have the fix. Your options would be to:

 * Pull down the latest version from farmbot/master, re-install dependencies and re-flash your device with the "edge" version of FarmBot OS (requires building from source).
 * Wait until version 4.0.0 (and a pre-compiled FarmBot OS image) come out, probably a week or so away.

**Comment by Klimbim1** [comment]
Hi @RickCarlino;
thanks very much for the support. How can I see which OS version works together with which Web-API version?

Cheers
Klim

**Comment by RickCarlino** [comment]
@Klimbim1 

 * Please see [FarmBot OS build instructions](https://github.com/FarmBot/farmbot_os/blob/master/docs/BUILDING.md) to learn how to compile FarmBot OS from source. Raise issues there if you require further assistance with building.
 * With regards to maintaining version compatibility: Your best bet is to simply use the latest version of each of the packages (Latest FBOS, latest MQTT, latest FE, etc). I would not recommend running older versions if possible. If you still wish to do this, you may use [git reset](https://git-scm.com/blog/2011/07/11/reset.html) to reset all repos to a particular date, but this is recommended only if you have a good understanding of Git.

We will begin tagging releases again with the upcoming 4.0.0 release.
